README
= ReadMe

Scout by Highgroove Studios

== Description

The easier way to monitor servers and web applications.

Scout makes monitoring and reporting on your web applications
as flexible and simple as possible.

Scout is a product of Highgroove Studios.
Please visit http://scoutapp.com for more information.

== Installing

Install the Scout gem:

  $ sudo gem install scout

Then simply run:

  $ scout

to run the installation wizard.  You'll need your server key to continue. Get the server key from your account at http://scoutapp.com

== Running the Scout Client

The Scout client has several modes of operation and commands.  The normal, intended usage is through a scheduled interval with no output.

Normal checkin with server:
  $ scout [OPTIONS] SERVER_KEY
  ... OR ...
  $ scout [OPTIONS] run SERVER_KEY

Install:
  $ scout
  ... OR ...
  $ scout [OPTIONS] install

Local plugin testing:
  $ scout [OPTIONS] test PATH_TO_PLUGIN [PLUGIN_OPTIONS]


SERVER_KEY is the identification key assigned by your account at http://scoutapp.com

PATH_TO_PLUGIN is the file system path to a Ruby file
that contains a Scout plugin.

PLUGIN_OPTIONS are one or more options in the form:
key1=val1 key2=val2
These options will be used for the plugin run.


== Setting up in cron

Configure Scout to run every minute. Typically, this will look like:

  * * * * *  deploy /usr/bin/scout SERVER_KEY

It's often helpful to log the output to a file. To do so:

  * * * * *  deploy /usr/bin/scout SERVER_KEY > /path/to/anywhere/scout.out 2>&1


For additional help, please visit http://scoutapp.com


CHANGELOG
== 5.0.3

* fixed regression: Error when running `scout AGENT_KEY` without first running scout and manually entering the agent key 

== 5.0.2

* fixed silent failure when plugin didn't inherit from Scout::Plugin
* beefed up error reporting

== 5.0.1

* plugin errors are now reported to scout server as errors, for easier plugin troubleshooting

== 5.0.0

* crontab must now run Scout every minute, regardless of what plan you are on
* Support for server downtime notifications
* Pings server every minute. Performs actual checkin on schedule provided by server
* Support for plugin option definition via an inline YAML file.
* Easier format for providing plugin arguments in test mode (scout help for details)
* Prints plugin arguments, including defaults, when run in test mode

== 4.0.2

* Check-in once after all plugins are run instead of once for each plugin for
  better efficiency

== 4.0.1

* Fixed a regression that broken support for some very old plugins

== 4.0.0

* Switched to the new API URL's
* Converted to JSON (using the vendored json_pure) from Marshal
* Upgraded to the data protocol used by the scout_agent
* Added SSL certificate verification to increase security
* Honor Last-Modified headers from the server to improve efficiency
* Added support for individual plugin timeouts
* Inserted a KILL signal for old processes to keep things running
* Removing obsolete clone action
* Removed non-functional test code
* Cleaned up Rake tasks for development
* Started sending an HTTP_CLIENT_HOSTNAME header to the Scout server
* Changed history file storage to be by plugin ID, instead of name

== 2.0.7

* Improved PID file error messages
* Adding a redundant Timeout to work around Net::HTTP hangs

== 2.0.6

* Adding plugin dependency support via the new needs() class method
* Improved Scout error backtraces (patch from dougbarth)

== 2.0.5

* Another Version bump to update gem servers

== 2.0.4

* Version bump to update gem servers

== 2.0.3

* Added documentation for Scout#data_for_server method for new plugin creation
* Added Version option for printing the current version
* Removed a spurious "puts" debug statement

== 2.0.2

* Fixed the logging bug I introduced by moving the PID check into the Command
  class

== 2.0.1

* Added some safety code to ensure SystemExit exceptions are not caught in our
  rescue clauses

== 2.0.0

* Reworked scout executable to work off an underlying command structure, similar
  to Subversion (a bare call and a call with just the key are supported for
  backward compatibility)
* Added various helper functions to Scout::Plugin to ease development
* Added a client clone command for instant setups

== 1.1.8

* Rectifying missing checkin -- this unifies 1.1.6 and 1.1.7 changes to gem

== 1.1.7

* Introducing a delta for the plugin run interval, now allowing runs even if
  they are up to 30 seconds early

== 1.1.6

* minor documentation update in scout installation wizard

== 1.1.5

* A more robust solution for plugin removal
* Added seconds to logging

== 1.1.4

* Trim all space from the client key during install
* Trying a fix for the plugin removal errors

== 1.1.3

* Fixed bug with running plugin using the -p option, new ensure wasn't returning
  the data

== 1.1.2

* Fixed the plugin interval not running on time due to a ">=" bug

== 1.1.1

* Fixed the double plugin load bug
* Ensuring that plugins are unloaded, even on error

== 1.1.0

* Using better url.path + url.query if present to properly encode URLs

== 1.0.9

* Fixed bug when plugin code would not compile, throws Exception
* Added ability to test/call scout on non-https servers (for debugging)
* Client now sends client version to server
* Client can send single values (using report, alert, error symbols) or multiple
  values (using reports, alerts, errors symbols)
* Added test suite, which is now the default rake task

== 1.0.8

* Added optional report field scout_time
* Changed #error method name to #scout_error to fix conflict with Logger#error

== 1.0.7

* Increased the plugin timeout to 60 seconds
* Added PID file protection to the client so only one copy will run at a time
* Fixed a bug that caused the wrong error message to be shown for the case when
  a plugin times out

== 1.0.6

* Improved error backtrace for local testing

== 1.0.5

* Added more documentation to Server and Plugin classes
* Fixed an issue where expand_path(~) would not work if HOME was not set, which  
  should help Scout run in OS X's LaunchDaemon using launchd

== 1.0.4

* Enhanced the -o option to take a Ruby Hash
* Fixed an issue where a failed plugin run would cause the client to skip all
  other plugins

== 1.0.3

* Refactored to allow testing of plugins locally using -p or --plugin option
  and -o or --plugin-options option

== 1.0.2

* Updated to use SSL by default for all communication to scout server
* Added elif dependency

== 1.0.1

* Fixed bug relating to history file – not using specified history file path

== 1.0.0

* Initial release


lib/scout/server.rb
#!/usr/bin/env ruby -wKU

require "net/https"
require "uri"
require "yaml"
require "timeout"
require "stringio"
require "zlib"
require "socket"

$LOAD_PATH << File.join(File.dirname(__FILE__), *%w[.. .. vendor json_pure lib])
require "json"

module Scout
  class Server
    # A new class for plugin Timeout errors.
    class PluginTimeoutError < RuntimeError; end
    # A new class for API Timeout errors.
    class APITimeoutError < RuntimeError; end
    
    # Headers passed up with all API requests.
    HTTP_HEADERS = { "CLIENT_VERSION"  => Scout::VERSION,
                     "CLIENT_HOSTNAME" => Socket.gethostname,
                     "ACCEPT_ENCODING" => "gzip" }
    
    # 
    # A plugin cannot take more than DEFAULT_PLUGIN_TIMEOUT seconds to execute, 
    # otherwise, a timeout error is generated.  This can be overriden by
    # individual plugins.
    # 
    DEFAULT_PLUGIN_TIMEOUT = 60
    #
    # A fuzzy range of seconds in which it is okay to rerun a plugin.
    # We consider the interval close enough at this point.
    # 
    RUN_DELTA = 30

    attr_reader :new_plan
    attr_reader :directives

    # Creates a new Scout Server connection.
    def initialize(server, client_key, history_file, logger = nil)
      @server       = server
      @client_key   = client_key
      @history_file = history_file
      @history      = Hash.new
      @logger       = logger
      @plugin_plan  = []
      @directives   = {} # take_snapshots
      @new_plan     = false

      # the block is only passed for install and test, since we split plan retrieval outside the lockfile for run
      if block_given?
        load_history
        yield self
        save_history
      end
    end

    #
    # Retrieves the Plugin Plan from the server. This is the list of plugins
    # to execute, along with all options.
    #
    # This method has a couple of side effects:
    # 1) it sets the @plugin plan with either A) whatever is in history, B) the results of the /plan retrieval
    # 2) it sets @checkin_to = true IF so directed by the scout server
    def fetch_plan
      url = urlify(:plan)
      info "Pinging server at #{url}..."
      headers = Hash.new
      if @history["plan_last_modified"] and @history["old_plugins"]
        headers["If-Modified-Since"] = @history["plan_last_modified"]
      end
      get(url, "Could not retrieve plan from server.", headers) do |res|
        if res.is_a? Net::HTTPNotModified
          info "Plan not modified. Will reuse saved plan."
          @plugin_plan = Array(@history["old_plugins"])
          @directives = @history["directives"] || Hash.new
        else
          info "plan has been modified. Will run the new plan now."
          begin
            body = res.body
            if res["Content-Encoding"] == "gzip" and body and not body.empty?
              body = Zlib::GzipReader.new(StringIO.new(body)).read
            end

            body_as_hash = JSON.parse(body)
            @plugin_plan = Array(body_as_hash["plugins"])
            @directives = body_as_hash["directives"].is_a?(Hash) ? body_as_hash["directives"] : Hash.new

            @history["plan_last_modified"] = res["last-modified"]
            @history["old_plugins"]        = @plugin_plan
            @history["directives"]         = @directives

            info "Plan loaded.  (#{@plugin_plan.size} plugins:  " +
                 "#{@plugin_plan.map { |p| p['name'] }.join(', ')})" +
                 ". Directives: #{@directives.to_a.map{|a|  "#{a.first}:#{a.last}"}.join(", ")}"

            @new_plan = true # used in determination if we should checkin this time or not
          rescue Exception
            fatal "Plan from server was malformed."
            exit
          end
        end
      end
    end

    # uses values from history and current time to determine if we should checkin at this time
    def time_to_checkin?
      @history['last_checkin'] == nil ||
              @directives['interval'] == nil ||
              (Time.now.to_i - Time.at(@history['last_checkin']).to_i).abs+15 > @directives['interval'].to_i*60
    rescue
      debug "Failed to calculate time_to_checkin. @history['last_checkin']=#{@history['last_checkin']}. "+
              "@directives['interval']=#{@directives['interval']}. Time.now.to_i=#{Time.now.to_i}"
      return true
    end

    # uses values from history and current time to determine if we should ping the server at this time
    def time_to_ping?
      return true if
      @history['last_ping'] == nil ||
              @directives['ping_interval'] == nil ||
              (Time.now.to_i - Time.at(@history['last_ping']).to_i).abs+15 > @directives['ping_interval'].to_i*60
    rescue
      debug "Failed to calculate time_to_ping. @history['last_ping']=#{@history['last_ping']}. "+
              "@directives['ping_interval']=#{@directives['ping_interval']}. Time.now.to_i=#{Time.now.to_i}"
      return true
    end

    # returns a human-readable representation of the next checkin, i.e., 5min 30sec
    def next_checkin
      secs= @directives['interval'].to_i*60 - (Time.now.to_i - Time.at(@history['last_checkin']).to_i).abs
      minutes=(secs.to_f/60).floor
      secs=secs%60
      "#{minutes}min #{secs} sec"
    rescue
      "[next scout invocation]"
    end

    # Runs all plugins from the given plan. Calls process_plugin on each plugin.
    # @plugin_execution_plan is populated by calling fetch_plan
    def run_plugins_by_plan
      prepare_checkin
      @plugin_plan.each do |plugin|
        begin
          process_plugin(plugin)
        rescue Exception
          @checkin[:errors] << build_report(
            plugin['id'],
            :subject => "Exception:  #{$!.message}.",
            :body    => $!.backtrace
          )
          error("Encountered an error: #{$!.message}")
        end
      end
      take_snapshot if @directives['take_snapshots']
      checkin
    end
    
    # 
    # This is the heart of Scout.  
    # 
    # First, it determines if a plugin is past interval and needs to be run.
    # If it is, it simply evals the code, compiling it.
    # It then loads the plugin and runs it with a PLUGIN_TIMEOUT time limit.
    # The plugin generates data, alerts, and errors. In addition, it will
    # set memory and last_run information in the history file.
    # 
    def process_plugin(plugin)
      info "Processing the '#{plugin['name']}' plugin:"
      id_and_name = "#{plugin['id']}-#{plugin['name']}".sub(/\A-/, "")
      last_run    = @history["last_runs"][id_and_name] ||
                    @history["last_runs"][plugin['name']]
      memory      = @history["memory"][id_and_name] ||
                    @history["memory"][plugin['name']]
      run_time    = Time.now
      delta       = last_run.nil? ? nil : run_time -
                                          (last_run + plugin['interval'] * 60)
      if last_run.nil? or delta.between?(-RUN_DELTA, 0) or delta >= 0
        debug "Plugin is past interval and needs to be run.  " +
              "(last run:  #{last_run || 'nil'})"
        debug "Compiling plugin..."
        begin
          eval( plugin['code'],
                TOPLEVEL_BINDING,
                plugin['path'] || plugin['name'] )
          info "Plugin compiled."
        rescue Exception
          raise if $!.is_a? SystemExit
          error "Plugin would not compile: #{$!.message}"
          @checkin[:errors] << build_report(plugin['id'],:subject => "Plugin would not compile", :body=>$!.message)
          return
        end
        debug "Loading plugin..."
        if job = Plugin.last_defined.load( last_run, (memory || Hash.new),
                                           plugin['options'] || Hash.new )
          info "Plugin loaded."
          debug "Running plugin..."
          begin
            data    = {}
            timeout = plugin['timeout'].to_i
            timeout = DEFAULT_PLUGIN_TIMEOUT unless timeout > 0
            Timeout.timeout(timeout, PluginTimeoutError) do
              data = job.run
            end
          rescue Timeout::Error, PluginTimeoutError
            error "Plugin took too long to run."
            @checkin[:errors] << build_report(plugin['id'],
                                              :subject => "Plugin took too long to run",
                                              :body=>"Execution timed out.")
            return
          rescue Exception
            raise if $!.is_a? SystemExit
            error "Plugin failed to run: #{$!.class}: #{$!.message}\n" +
                  "#{$!.backtrace.join("\n")}"
            @checkin[:errors] << build_report(plugin['id'],
                                              :subject => "Plugin failed to run",
                                              :body=>"#{$!.class}: #{$!.message}\n#{$!.backtrace.join("\n")}")
          end
          info "Plugin completed its run."
          
          %w[report alert error summary].each do |type|
            plural  = "#{type}s".sub(/ys\z/, "ies").to_sym
            reports = data[plural].is_a?(Array) ? data[plural] :
                                                  [data[plural]].compact
            if report = data[type.to_sym]
              reports << report
            end
            reports.each do |fields|
              @checkin[plural] << build_report(plugin['id'], fields)
            end
          end
          
          @history["last_runs"].delete(plugin['name'])
          @history["memory"].delete(plugin['name'])
          @history["last_runs"][id_and_name] = run_time
          @history["memory"][id_and_name]    = data[:memory]
        else
          @checkin[:errors] << build_report(
            plugin['id'],
            :subject => "Plugin would not load."
          )
        end
      else
        debug "Plugin does not need to be run at this time.  " +
              "(last run:  #{last_run || 'nil'})"
      end
      data
    ensure
      if Plugin.last_defined
        debug "Removing plugin code..."
        begin
          Object.send(:remove_const, Plugin.last_defined.to_s.split("::").first)
          Plugin.last_defined = nil
          info "Plugin Removed."
        rescue
          raise if $!.is_a? SystemExit
          error "Unable to remove plugin."
        end
      end
      info "Plugin '#{plugin['name']}' processing complete."
    end


    # captures a list of processes running at this moment
    def take_snapshot
      info "Taking a process snapshot"
      ps=%x(ps aux).split("\n")[1..-1].join("\n") # get rid of the header line
      @checkin[:snapshot]=ps
      rescue Exception
        error "unable to capture processes on this server. #{$!.message}"
        return nil
    end

    # Prepares a check-in data structure to hold Plugin generated data.
    def prepare_checkin
      @checkin = { :reports   => Array.new,
                   :alerts    => Array.new,
                   :errors    => Array.new,
                   :summaries => Array.new,
                   :snapshot  => '' }
    end

    def show_checkin(printer = :p)
      send(printer, @checkin)
    end

    #
    # Loads the history file from disk. If the file does not exist,
    # it creates one.
    #
    def load_history
      unless File.exist? @history_file
        create_blank_history
      end
      debug "Loading history file..."
      @history = File.open(@history_file) { |file| YAML.load(file) }
      info "History file loaded."
    end

    # creates a blank history file
    def create_blank_history
      debug "Creating empty history file..."
      File.open(@history_file, "w") do |file|
        YAML.dump({"last_runs" => Hash.new, "memory" => Hash.new}, file)
      end
      info "History file created."      
    end    

    # Saves the history file to disk.
    def save_history
      debug "Saving history file..."
      File.open(@history_file, "w") { |file| YAML.dump(@history, file) }
      info "History file saved."
    end

    private
    
    def build_report(plugin_id, fields)
      { :plugin_id  => plugin_id,
        :created_at => Time.now.utc.strftime("%Y-%m-%d %H:%M:%S"),
        :fields     => fields }
    end

    def urlify(url_name, options = Hash.new)
      return unless @server
      options.merge!(:client_version => Scout::VERSION)
      URI.join( @server,
                "/clients/CLIENT_KEY/#{url_name}.scout".
                  gsub(/\bCLIENT_KEY\b/, @client_key).
                  gsub(/\b[A-Z_]+\b/) { |k| options[k.downcase.to_sym] || k } )
    end
    
    def post(url, error, body, headers = Hash.new, &response_handler)
      return unless url
      request(url, response_handler, error) do |connection|
        post = Net::HTTP::Post.new( url.path +
                                    (url.query ? ('?' + url.query) : ''),
                                    HTTP_HEADERS.merge(headers) )
        post.body = body
        connection.request(post)
      end
    end

    def get(url, error, headers = Hash.new, &response_handler)
      return unless url
      request(url, response_handler, error) do |connection|
        connection.get( url.path + (url.query ? ('?' + url.query) : ''),
                        HTTP_HEADERS.merge(headers) )
      end
    end
    
    def request(url, response_handler, error, &connector)
      response           = nil
      Timeout.timeout(5 * 60, APITimeoutError) do
        http               = Net::HTTP.new(url.host, url.port)
        if url.is_a? URI::HTTPS
          http.use_ssl     = true
          http.ca_file     = File.join( File.dirname(__FILE__),
                                        *%w[.. .. data cacert.pem] )
          http.verify_mode = OpenSSL::SSL::VERIFY_PEER |
                             OpenSSL::SSL::VERIFY_FAIL_IF_NO_PEER_CERT
        end
        response           = no_warnings { http.start(&connector) }
      end
      case response
      when Net::HTTPSuccess, Net::HTTPNotModified
        response_handler[response] unless response_handler.nil?
      else
        error = "Server says: #{response['x-scout-msg']}" if response['x-scout-msg']
        fatal error
        raise SystemExit.new(error)
      end
    rescue Timeout::Error
      fatal "Request timed out."
      exit
    rescue Exception
      raise if $!.is_a? SystemExit
      fatal "An HTTP error occurred:  #{$!.message}"
      exit
    end
    
    def checkin
      @history['last_checkin'] = Time.now.to_i # might have to save the time of invocation and use here to prevent drift
      io   =  StringIO.new
      gzip =  Zlib::GzipWriter.new(io)
      gzip << @checkin.to_json
      gzip.close
      post( urlify(:checkin),
            "Unable to check in with the server.",
            io.string,
            "Content-Type"     => "application/json",
            "CONTENT_ENCODING" => "gzip" )
    rescue Exception
      error "Unable to check in with the server."
    end
    
    
    def no_warnings
      old_verbose = $VERBOSE
      $VERBOSE    = false
      yield
    ensure
      $VERBOSE = old_verbose
    end
    
    # Forward Logger methods to an active instance, when there is one.
    def method_missing(meth, *args, &block)
      if (Logger::SEV_LABEL - %w[ANY]).include? meth.to_s.upcase
        @logger.send(meth, *args, &block) unless @logger.nil?
      else
        super
      end
    end
  end
end
